ext/ddtrace_profiling_native_extension/ruby_helpers.c

```diff
@@ -210,3 +214,44 @@ size_t ruby_obj_memsize_of(VALUE obj) {
       return 0;
   }
 }
+
+// Inspired by rb_class_of but without actually returning classes or potentially doing assertions
+static bool ruby_is_obj_with_class(VALUE obj) {
+  if (!RB_SPECIAL_CONST_P(obj)) {
+    return true;
+  }
+  if (obj == RUBY_Qfalse) {
+    return true;
+  }
+  else if (obj == RUBY_Qnil) {
+    return true;
+  }
+  else if (obj == RUBY_Qtrue) {
+    return true;
+  }
+  else if (RB_FIXNUM_P(obj)) {
+    return true;
+  }
+  else if (RB_STATIC_SYM_P(obj)) {
+    return true;
+  }
+  else if (RB_FLONUM_P(obj)) {
+    return true;
+  }
+
+  return false;
+}
+
+VALUE ruby_safe_inspect(VALUE obj) {
+  if (!ruby_is_obj_with_class(obj)) {
+    return Qnil;
+  }
+
+  if (rb_respond_to(obj, inspect_id)) {
+    return rb_sprintf("%+"PRIsVALUE, obj);
+  } else if (rb_respond_to(obj, to_s_id)) {
+    return rb_sprintf("%"PRIsVALUE, obj);
+  } else {
+    return Qnil;
+  }
+}
```

Comment:
Minor: Since this is for debugging it may be worth making explicit the weird cases?

```suggestion
VALUE ruby_safe_inspect(VALUE obj) {
  if (!ruby_is_obj_with_class(obj)) {
    return rb_str_new_cstr("(Not an object)");
  }

  if (rb_respond_to(obj, inspect_id)) {
    return rb_sprintf("%+"PRIsVALUE, obj);
  } else if (rb_respond_to(obj, to_s_id)) {
    return rb_sprintf("%"PRIsVALUE, obj);
  } else {
    return rb_str_new_cstr("(Not inspectable)");
  }
}
```